ddtrace/contrib/internal/google_generativeai/_utils.py

```diff
@@ -0,0 +1,217 @@
+import sys
+
+import wrapt
+
+from ddtrace.internal.utils import get_argument_value
+
+
+class BaseTracedGenerateContentResponse(wrapt.ObjectProxy):
+    """Base wrapper class for GenerateContentResponse objects for tracing streamed responses."""
+
+    def __init__(self, wrapped, instance, integration, span, args, kwargs):
+        super().__init__(wrapped)
+        self._model_instance = instance
+        self._dd_integration = integration
+        self._dd_span = span
+        self._args = args
+        self._kwargs = kwargs
+
+
+class TracedGenerateContentResponse(BaseTracedGenerateContentResponse):
+    def __iter__(self):
+        try:
+            for chunk in self.__wrapped__.__iter__():
+                yield chunk
+        except Exception:
+            self._dd_span.set_exc_info(*sys.exc_info())
+            raise
+        else:
+            tag_response(self._dd_span, self.__wrapped__, self._dd_integration, self._model_instance)
+        finally:
+            self._dd_span.finish()
+
+
+class TracedAsyncGenerateContentResponse(BaseTracedGenerateContentResponse):
+    async def __aiter__(self):
+        try:
+            async for chunk in self.__wrapped__.__aiter__():
+                yield chunk
+        except Exception:
+            self._dd_span.set_exc_info(*sys.exc_info())
+            raise
+        else:
+            tag_response(self._dd_span, self.__wrapped__, self._dd_integration, self._model_instance)
+        finally:
+            self._dd_span.finish()
+
+
+def _extract_model_name(instance):
+    """Extract the model name from the instance."""
+    model_name = getattr(instance, "model_name", "")
+    if not model_name or not isinstance(model_name, str):
+        return ""
+    if "/" in model_name:
+        return model_name.split("/")[-1]
+    return model_name
+
+
+def _extract_api_key(instance):
+    """Extract the API key from the model instance."""
+    client = getattr(instance, "_client", None)
+    if getattr(instance, "_async_client", None):
+        client = getattr(instance._async_client, "_client", None)
+    if not client:
+        return None
+    client_options = getattr(client, "_client_options", None)
+    if not client_options:
+        return None
+    return getattr(client_options, "api_key", None)
+
+
+def _tag_request_content_part(span, integration, part, part_idx, content_idx):
+    """Tag the generation span with request content parts."""
+    text = getattr(part, "text", "")
+    function_call = getattr(part, "function_call", None)
+    function_response = getattr(part, "function_response", None)
+    if isinstance(part, dict):
+        text = part.get("text", "")
+        function_call = part.get("function_call", None)
+        function_response = part.get("function_response", None)
```

Comment:
```suggestion
    if isinstance(part, dict):
        text = part.get("text", "")
        function_call = part.get("function_call", None)
        function_response = part.get("function_response", None)
    else:
        text = getattr(part, "text", "")
        function_call = getattr(part, "function_call", None)
        function_response = getattr(part, "function_response", None)
```

Not 100% sure but isn't it better this way ? we would only execute the getattr if the part is not a dict.
Also wouldn't getattr on a dict return an attribute not found exception that we aren't catching ?

Reply:
Nvm i just noticed we have a default None or empty string for the case where it's not found.

Reply:
Will be adding a convenience method to extract this instead!